apps/prairielearn/src/lib/base64-util.test.ts

```diff
@@ -0,0 +1,90 @@
+import { describe, expect, it } from 'vitest';
+
+import { run } from '@prairielearn/run';
+
+import {
+  b64DecodeUnicode,
+  b64DecodeUnicodeBrowser,
+  b64DecodeUnicodeNode,
+  b64EncodeUnicode,
+  b64EncodeUnicodeBrowser,
+  b64EncodeUnicodeNode,
+} from './base64-util.js';
+
+describe('base64 encoding/decoding equivalence', () => {
+  const testCases: { name: string; value: string }[] = [
+    // ASCII cases
+    { name: 'simple ASCII string', value: 'hello' },
+    { name: 'string with punctuation', value: 'Hello, World!' },
+    { name: 'numeric string', value: '123456789' },
+
+    // Whitespace and escapes
+    { name: 'whitespace characters', value: '\n\t\r' },
+    { name: 'escaped whitespace strings', value: '\\n\\t\\r' },
+
+    // Symbols and specials
+    { name: 'special characters', value: 'special!@#$%^&*()' },
```

Comment:
Nit: I'd suggest adding characters that have special meaning in HTML, JS or URLs, for extra precaution.

```suggestion
    { name: 'special characters', value: 'special!@#$%^&*()<>?"\'`=+\\' },
```